kernel/src/log_segment.rs

```diff
@@ -213,17 +218,141 @@ impl LogSegment {
+                checkpoint_read_schema,
+                meta_predicate,
+            )?
+        } else {
+            engine.get_parquet_handler().read_parquet_files(
+                &checkpoint_file_meta,
+                checkpoint_read_schema,
+                meta_predicate,
+            )?
+        };
+
+        let parquet_handler = engine.get_parquet_handler();
+        let actions_iter = actions
+            .map(move |batch_result| -> DeltaResult<_> {
+                let checkpoint_batch = batch_result?;
+                // This closure maps the checkpoint batch to an iterator of batches
+                // by chaining the checkpoint batch with sidecar batches if they exist.
+
+                // 1. In the case where the schema does not contain add/remove actions, we return the checkpoint
+                // batch directly as sidecar files only have to be read when the schema contains add/remove actions.
+                // 2. Multi-part checkpoint batches never have sidecar actions, so the batch is returned as-is.
+                let sidecar_content = if !need_add_actions || checkpoint_parts.len() > 1 {
+                    None
+                } else {
+                    Self::process_sidecars(
+                        parquet_handler.clone(), // cheap Arc clone
+                        log_root.clone(),
+                        checkpoint_batch.as_ref(),
+                    )?
+                };
+
+                let combined_batches = std::iter::once(Ok(checkpoint_batch))
+                    .chain(sidecar_content.into_iter().flatten())
+                    .map_ok(|sidecar_batch| (sidecar_batch, false));
+
+                Ok(combined_batches)
+            })
+            .flatten_ok()
+            // Map converts Result<Result<Box<dyn EngineData>, _>,_> to Result<Box<dyn EngineData>, _>
+            .map(|result| result?);
+
+        Ok(actions_iter)
+    }
+
+    /// Processes sidecar files for the given checkpoint batch.
+    ///
+    /// This function extracts any sidecar file references from the provided batch.
+    /// Each sidecar file is read and an iterator of sidecar file batches is returned
+    fn process_sidecars(
+        parquet_handler: Arc<dyn ParquetHandler>,
+        log_root: Url,
+        batch: &dyn EngineData,
+    ) -> DeltaResult<Option<impl Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send>> {
+        // Visit the rows of the checkpoint batch to extract sidecar file references
+        let mut visitor = SidecarVisitor::default();
+        visitor.visit_rows_of(batch)?;
+
+        // If there are no sidecar files, return early
+        if visitor.sidecars.is_empty() {
+            return Ok(None);
+        }
+
+        let sidecar_files: Vec<_> = visitor
+            .sidecars
+            .iter()
+            .map(|sidecar| Self::sidecar_to_filemeta(sidecar, &log_root))
+            .try_collect()?;
+
+        let sidecar_read_schema = get_log_add_schema().clone();
```

Comment:
not: you don't need this as a variable, just do `get_log_add_schema().clone()` as the arg to `read_parquet_files` below